File: solutions/julia/all-your-base/1/all-your-base.jl
function validate_input(digits, base_in, base_out)
    if base_in<=1
        throw(DomainError(base_in, "Invalid base"))
    elseif base_out<=1
        throw(DomainError(base_out, "Invalid base"))
    else
        for i in eachindex(digits)
            if digits[i]<0 || digits[i]>=base_in 
                throw(DomainError(digits[i], "Invalid Digit"))
            end
        end        
    end
end

function digits_to_int(digits, base)
    num = 0
    b_power = 1
    ds = reverse(digits)
    for i in eachindex(ds)
        num += ds[i]*b_power
        b_power *= base
    end
    num
end

function int_to_digits(num, base)
    digits = []
    while num>0
        push!(digits, num % base)
        num ÷= base
    end
    reverse(digits)
end


function all_your_base(digits, base_in, base_out)
    validate_input(digits, base_in, base_out)
    number = digits_to_int(digits, base_in)
    result = int_to_digits(number, base_out)    
    isempty(result) ? [0] : result
end

File: solutions/cpp/all-your-base/1/all_your_base.cpp
#include "all_your_base.h"
#include <vector>
#include <cstddef>
#include <stdexcept>
#include <stdlib.h>

using namespace std;

namespace all_your_base {

bool is_valid (int a, vector<unsigned int> digits){
      for(size_t i=0; i!=digits.size(); ++i){
              if (digits.at(i)>=abs(a))
                  return false;
            }  
        return true;
}

int digits_to_int(int a, vector<unsigned int> digits){
    int num = 0;
    int a_power = 1;
    for(size_t i=digits.size(); i--;){
        num +=digits.at(i)*a_power;
        a_power *= a;
    }
    return num;
}

vector<unsigned int> int_to_digits(int num, int b){
    vector<unsigned int> digits;
    while(num>0){
        digits.insert(digits.begin(),(num % b));
        num = num/b;
    }
    return digits;
}

vector<unsigned int> convert(int a, vector<unsigned int> digits, int b){
    if(((a>1) && (b>1)) && (is_valid(a, digits))){
         int number = digits_to_int(a, digits);
         return int_to_digits(number, b);
    } else 
        throw std::invalid_argument("Invalid input.");
 }

}  // namespace all_your_base
